File: duplex/java/src/main/java/io/nightseam/duplex/Dispatcher.java
package io.nightseam.duplex;

import dev.bitspark.bitwire.*;
import java.util.*;

/** Explicit path routing over one borrowed Bitwire endpoint attachment. */
public final class Dispatcher implements Wire, AutoCloseable {
    private record Route(List<String> path, boolean prefix) {}
    private record Registration(Route route, Receiver receiver) {}
    private final Endpoint root;
    private final Map<Route,Registration> routes = new LinkedHashMap<>();
    private final WeakHashMap<ReturnAddress,Map<String,Registration>> calls = new WeakHashMap<>();
    private Runnable detach;
    private boolean ended;

    public Dispatcher(Endpoint root) {
        this.root = Objects.requireNonNull(root);
        Runnable acquired = root.receive(new Receiver(this::dispatch, this::close));
        synchronized (this) { if (!ended) { detach = acquired; return; } }
        acquired.run();
        throw new IllegalStateException("endpoint closed during attachment");
    }
    @Override public void send(List<String> path, Message message) {
        synchronized (this) { requireOpen(); }
        root.send(path, message);
    }
    public Runnable register(List<String> path, Receiver receiver) { return register(path, false, receiver); }
    public Runnable registerPrefix(List<String> path, Receiver receiver) { return register(path, true, receiver); }
    private synchronized Runnable register(List<String> path, boolean prefix, Receiver receiver) {
        requireOpen(); Wires.encodePath(path);
        var route = new Route(List.copyOf(path), prefix);
        var registration = new Registration(route, receiver);
        if (routes.putIfAbsent(route, registration) != null) throw new IllegalStateException("route already registered");
        return () -> { synchronized (Dispatcher.this) { routes.remove(route, registration); } };
    }
    private void dispatch(List<String> path, Message message) {
        Registration selected = null;
        synchronized (this) {
            if (ended) return;
            if (message.frame() instanceof ProfileFrame.Cancel cancel && message.returnAddress() != null) {
                var captured = calls.get(message.returnAddress());
                if (captured != null) selected = captured.remove(cancel.id());
            } else {
                selected = routes.get(new Route(path, false));
                if (selected == null) for (var candidate : routes.values()) {
                    var prefix = candidate.route().path();
                    if (candidate.route().prefix() && path.size() >= prefix.size()
                        && path.subList(0, prefix.size()).equals(prefix)
                        && (selected == null || prefix.size() > selected.route().path().size())) selected = candidate;
                }
                if (selected != null && message.frame() instanceof ProfileFrame.Request request && message.returnAddress() != null)
                    calls.computeIfAbsent(message.returnAddress(), key -> new HashMap<>()).put(request.id(), selected);
            }
        }
        if (selected != null && selected.receiver().message() != null) selected.receiver().message().accept(path, message);
        else if (message.frame() instanceof ProfileFrame.Request request && message.returnAddress() != null)
            message.returnAddress().wire().send(List.of(), new Message(new ProfileFrame.Response(request.id(),
                null, new ProfileError("method_not_found", "Unknown method"), request.traceparent(), request.tracestate()), null));
    }
    public Endpoint select(List<String> path) {
        Wires.encodePath(path);
        var prefix = List.copyOf(path);
        return new Endpoint() {
            private Object token;
            private Runnable release;
            private Receiver receiver;
            private boolean closed;
            @Override public void send(List<String> relative, Message message) {
                synchronized (this) { if (closed) throw new IllegalStateException("endpoint closed"); }
                var joined = new ArrayList<>(prefix); joined.addAll(relative);
                Dispatcher.this.send(joined, message);
            }
            @Override public synchronized Runnable receive(Receiver value) {
                if (closed || token != null) throw new IllegalStateException("endpoint closed or already attached");
                Object current = new Object(); token = current; receiver = value;
                try {
                    release = registerPrefix(prefix, new Receiver(value.message() == null ? null : (delivered, message) ->
                        value.message().accept(List.copyOf(delivered.subList(prefix.size(), delivered.size())), message), this::close));
                } catch (RuntimeException failure) { token = null; receiver = null; throw failure; }
                return () -> { synchronized (this) {
                    if (token != current) return;
                    token = null; receiver = null; Runnable held = release; release = null; held.run();
                } };
            }
            @Override public void close(int code, String reason) {
                Receiver held; Runnable off;
                synchronized (this) {
                    if (closed) return;
                    closed = true; token = null; held = receiver; receiver = null; off = release; release = null;
                }
                if (off != null) off.run();
                if (held != null && held.closed() != null) held.closed().accept(code, reason);
            }
        };
    }
    private void requireOpen() { if (ended) throw new IllegalStateException("dispatcher closed"); }
    public void close(int code, String reason) {
        List<Registration> held; Runnable off;
        synchronized (this) {
            if (ended) return;
            ended = true; held = List.copyOf(routes.values()); routes.clear(); calls.clear(); off = detach; detach = null;
        }
        if (off != null) off.run();
        for (var registration : held) if (registration.receiver().closed() != null) registration.receiver().closed().accept(code, reason);
    }
    @Override public void close() { close(1000, "done"); }
}

File: duplex/swift/NightseamDuplex/Sources/NightseamDuplex/Dispatcher.swift
import Bitwire
import Foundation

/// Explicit path dispatch over one attachment to a borrowed Bitwire endpoint.
public final class Dispatcher: Wire, @unchecked Sendable {
    private struct Route: Hashable { let path: Data; let prefix: Bool }
    private struct Registration: Sendable { let id: UUID; let path: [String]; let receiver: Receiver }
    private struct Captured {
        weak var address: ReturnAddress?
        var calls: [String: Registration]
    }
    private let root: any Endpoint
    private let lock = NSLock()
    private var routes: [Route: Registration] = [:]
    private var captured: [ObjectIdentifier: Captured] = [:]
    private var ended = false
    private var detach: Detach?

    public init(_ root: any Endpoint) throws {
        self.root = root
        let acquired = try root.receive(receiver: Receiver(message: { [weak self] path, message in
            self?.dispatch(path, message)
        }, closed: { [weak self] code, reason in self?.close(code: code, reason: reason) }))
        let active = lock.withLock { if ended { return false }; detach = acquired; return true }
        if !active { acquired(); throw DuplexError.closed }
    }
    deinit { detach?() }
    public func send(path: [String], message: Message) throws {
        guard lock.withLock({ !ended }) else { throw DuplexError.closed }
        try root.send(path: path, message: message)
    }
    public func register(path: [String], receiver: Receiver) throws -> Detach {
        try register(path: path, prefix: false, receiver: receiver)
    }
    public func registerPrefix(path: [String], receiver: Receiver) throws -> Detach {
        try register(path: path, prefix: true, receiver: receiver)
    }
    private func register(path: [String], prefix: Bool, receiver: Receiver) throws -> Detach {
        let route = Route(path: Path.key(path), prefix: prefix)
        let registration = Registration(id: UUID(), path: path, receiver: receiver)
        try lock.withLock {
            guard !ended else { throw DuplexError.closed }
            guard routes[route] == nil else { throw DuplexError.receiverExists }
            routes[route] = registration
        }
        return { [weak self] in self?.lock.withLock {
            if self?.routes[route]?.id == registration.id { self?.routes.removeValue(forKey: route) }
        } }
    }
    private func dispatch(_ path: [String], _ message: Message) {
        let selected: Registration? = lock.withLock {
            guard !ended else { return nil }
            captured = captured.filter { $0.value.address != nil }
            if message.frame.kind == .cancel, let address = message.returnAddress, let id = message.frame.id {
                return captured[ObjectIdentifier(address)]?.calls.removeValue(forKey: id)
            }
            var selected = routes[Route(path: Path.key(path), prefix: false)]
            if selected == nil {
                for (route, candidate) in routes where route.prefix && Path.hasPrefix(path, candidate.path) {
                    if selected == nil || candidate.path.count > selected!.path.count { selected = candidate }
                }
            }
            if let selected, message.frame.kind == .request, let address = message.returnAddress, let id = message.frame.id {
                let key = ObjectIdentifier(address)
                if captured[key] == nil { captured[key] = Captured(address: address, calls: [:]) }
                captured[key]?.calls[id] = selected
            }
            return selected
        }
        if let selected { selected.receiver.message(path, message) }
        else if message.frame.kind == .request, let address = message.returnAddress {
            try? address.wire.send(path: [], message: Message(frame: ProfileFrame(kind: .response, id: message.frame.id,
                error: ProfileError(code: "method_not_found", message: "Unknown method"),
                traceparent: message.frame.traceparent, tracestate: message.frame.tracestate)))
        }
    }
    public func select(path: [String]) -> any Endpoint { RoutedEndpoint(owner: self, path: path) }
    public func close(code: Int = 1000, reason: String = "done") {
        let held: ([Registration], Detach?) = lock.withLock {
            guard !ended else { return ([], nil) }
            ended = true
            let held = (Array(routes.values), detach)
            routes.removeAll(); captured.removeAll(); detach = nil
            return held
        }
        held.1?()
        for registration in held.0 { registration.receiver.closed(code, reason) }
    }
}

private final class RoutedEndpoint: Endpoint, @unchecked Sendable {
    let owner: Dispatcher
    let path: [String]
    let lock = NSRecursiveLock()
    var ended = false
    var attachment: (UUID, Receiver, Detach)?
    init(owner: Dispatcher, path: [String]) { self.owner = owner; self.path = path }
    deinit { attachment?.2() }
    func send(path: [String], message: Message) throws {
        guard lock.withLock({ !ended }) else { throw DuplexError.closed }
        try owner.send(path: self.path + path, message: message)
    }
    func receive(receiver: Receiver) throws -> Detach {
        let id = UUID()
        try lock.withLock {
            guard !ended else { throw DuplexError.closed }
            guard attachment == nil else { throw DuplexError.receiverExists }
            let prefix = path
            let detach = try owner.registerPrefix(path: prefix, receiver: Receiver(message: { delivered, message in
                receiver.message(Array(delivered.dropFirst(prefix.count)), message)
            }, closed: { [weak self] code, reason in try? self?.close(code: code, reason: reason) }))
            attachment = (id, receiver, detach)
        }
        return { [weak self] in
            guard let self else { return }
            let detach: Detach? = lock.withLock {
                guard attachment?.0 == id else { return nil }
                let held = attachment?.2; attachment = nil; return held
            }
            detach?()
        }
    }
    func close(code: Int, reason: String) throws {
        let held: (UUID, Receiver, Detach)? = lock.withLock {
            guard !ended else { return nil }
            ended = true; let held = attachment; attachment = nil; return held
        }
        held?.2(); held?.1.closed(code, reason)
    }
}
